### PR TITLE
add e2e framework sample

### PR DESCRIPTION
**Pull request by zzxwill**

### CODE REVIEW

**Comment by zzxwill** [comment]
The build hit an issue 'create client from kubeconfig err Get http://localhost:8080/api?timeout=32s: dial tcp [::1]:8080: connect: connection refused'.

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -64,11 +65,13 @@ workloaddefinition.core.oam.dev/deployments.apps                      deployment
 workloaddefinition.core.oam.dev/statefulsets.apps                     statefulsets.apps
 ```
 
-## Test
+## E2E test
 ```
-$ go test ./pkg/test
-ok  	github.com/cloud-native-application/rudrx/pkg/test	14.662s
+$ cd pkg/test
+$ ginkgo -v -r
```
Comment (wonderflow) [inline DEVELOPMENT.md:71]: put this into Makefile `e2e-test` please

Comment (zzxwill) [reply DEVELOPMENT.md]: Done.

**Comment by wonderflow** [comment]
you can move `test/` dir outside of `pkg/`, then it won't trigger unit test to run

**Comment by zzxwill** [comment]
Maybe I need to run `rudr admin:init` first to fix the build issue.

**Comment by zzxwill** [comment]
add rudr env test cases and refine rudr run:workload
```
make e2e-test
# Run e2e test
ginkgo -v -r e2e
Running Suite: Workload Suite
=============================
Random Seed: 1596458182
Will run 4 of 4 specs

Workload env init
  should print env initiation successful message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/component/component_test.go:19
Create env succeed, current env is default
•
------------------------------
Workload env sw
  should show env switch message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/component/component_test.go:29
Switch env succeed, current env is default
•
------------------------------
Workload run
  should print successful creation information
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/component/component_test.go:39
Creating AppConfig app-testworkloadrun-basic
SUCCEED
•
------------------------------
Workload delete
  should print successful deletion information
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/component/component_test.go:48
Deleting AppConfig "app-testworkloadrun-basic"
DELETE SUCCEED
•
Ran 4 of 4 Specs in 10.376 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Running Suite: Env Suite
========================
Random Seed: 1596458182
Will run 7 of 7 specs

Env env init
  should print env initiation successful message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:18
Create env succeed, current env is env-hello
•
------------------------------
Env env init another one
  should print env initiation successful message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:28
Create env succeed, current env is env-world
•
------------------------------
Env env show
  should show detailed env message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:38
NAME     	NAMESPACE
env-hello	env-hello•
------------------------------
Env env sw
  should show env switch message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:50
Switch env succeed, current env is env-hello
•
------------------------------
Env env list
  should list all envs
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:60
NAME        	NAMESPACE
default     	default
env-hello   	env-hello
env-hello2  	env-hello2
env-workload	env-workload
env-world   	env-world
test        	test        •
------------------------------
Env env delete
  should delete all envs
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:71
env-world deleted
•
------------------------------
Env env delete currently using one
  should delete all envs
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/env/env_test.go:81
Error: you can't delete current using env env-hello
•
Ran 7 of 7 Specs in 9.247 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 2 suites in 22.338449566s
Test Suite Passed
```

**Comment by zzxwill** [comment]
Added trait attach cases and refactor all others
```
Running Suite: Trait Suite
==========================
Random Seed: 1596559178
Will run 5 of 5 specs

Trait env init
  should print env initiation successful message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/commonContext.go:14
Create env succeed, current env is default
•
------------------------------
Trait env switch
  should show env switch message
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/commonContext.go:40
Switch env succeed, current env is default
•
------------------------------
Trait run
  should print successful creation information
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/commonContext.go:76
Creating AppConfig app-trait-basic
SUCCEED
•
------------------------------
Trait rudr attach trait
  should print successful attached information
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/trait/trait_test.go:24
Applying trait for app
Succeeded!
•
------------------------------
Trait delete
  should print successful deletion information
  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX/e2e/commonContext.go:85
Deleting AppConfig "app-trait-basic"
DELETE SUCCEED
•
Ran 5 of 5 Specs in 9.717 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

**Comment by zzxwill** [comment]
Investigating the failure of e2e environment preparation. 
While in a clean Linux, kind cluster, OAM runtime and manifests can successfully applied.
```
[root@iZj6cc54ecpc3pm8t9yctuZ ~]# kind create cluster
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.18.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
[root@iZj6cc54ecpc3pm8t9yctuZ ~]# k get nodes -o wide
NAME                 STATUS     ROLES    AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION               CONTAINER-RUNTIME
kind-control-plane   NotReady   master   31s   v1.18.2   172.18.0.2    <none>        Ubuntu 19.10   3.10.0-957.21.3.el7.x86_64   containerd://1.3.3-14-g449e9269
[root@iZj6cc54ecpc3pm8t9yctuZ ~]# kubectl create namespace oam-system
namespace/oam-system created
[root@iZj6cc54ecpc3pm8t9yctuZ ~]#           helm repo add crossplane-master https://charts.crossplane.io/master/

"crossplane-master" has been added to your repositories
[root@iZj6cc54ecpc3pm8t9yctuZ ~]#           helm install oam --namespace oam-system crossplane-master/oam-kubernetes-runtime --devel

NAME: oam
LAST DEPLOYED: Thu Aug  6 19:41:27 2020
NAMESPACE: oam-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
[root@iZj6cc54ecpc3pm8t9yctuZ ~]#
[root@iZj6cc54ecpc3pm8t9yctuZ ~]# git clone https://github.com/zzxwill/Rudrx -b e2e
Cloning into 'Rudrx'...
remote: Enumerating objects: 84, done.
remote: Counting objects: 100% (84/84), done.
remote: Compressing objects: 100% (54/54), done.
remote: Total 876 (delta 30), reused 69 (delta 25), pack-reused 792
Receiving objects: 100% (876/876), 483.73 KiB | 282.00 KiB/s, done.
Resolving deltas: 100% (400/400), done.
[root@iZj6cc54ecpc3pm8t9yctuZ ~]# cd Rudrx/
[root@iZj6cc54ecpc3pm8t9yctuZ Rudrx]# kubectl apply -f .github/workflows/core.oam.dev_workloaddefinitions.yaml
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
customresourcedefinition.apiextensions.k8s.io/workloaddefinitions.core.oam.dev configured
[root@iZj6cc54ecpc3pm8t9yctuZ Rudrx]#           kubectl apply -f .github/workflows/core.oam.dev_traitdefinitions.yaml
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
customresourcedefinition.apiextensions.k8s.io/traitdefinitions.core.oam.dev configured
[root@iZj6cc54ecpc3pm8t9yctuZ Rudrx]# kubectl apply -R -f config/samples
traitdefinition.core.oam.dev/manualscalertrait.core.oam.dev created
traitdefinition.core.oam.dev/simplerollouttraits.extend.oam.dev created
workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev created
workloaddefinition.core.oam.dev/deployments.apps created


```
image

```

File: .github/workflows/core.oam.dev_traitdefinitions.yaml [hunk 2]
```diff
@@ -0,0 +1,87 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (wonderflow) [inline .github/workflows/core.oam.dev_traitdefinitions.yaml:1]: oam-k8s-runtime has fixed this issue, we can remove these definitions

File: e2e/application/application_test.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"fmt"
+
+	"github.com/cloud-native-application/rudrx/e2e"
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+)
+
+var (
+	//TODO(zzxwill) Need to change env name after [issue#82](https://github.com/cloud-native-application/RudrX/issues/82) is fixed.
```
Comment (wonderflow) [inline e2e/application/application_test.go:12]: it's fixed now

File: e2e/workload/workload_test.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"fmt"
+
+	"github.com/cloud-native-application/rudrx/e2e"
+	"github.com/onsi/ginkgo"
+)
+
+var (
+	//TODO(zzxwill) Need to change env name after [issue#82](https://github.com/cloud-native-application/RudrX/issues/82) is fixed.
```
Comment (wonderflow) [inline e2e/workload/workload_test.go:11]: fixed now

File: .github/workflows/go.yml [hunk 5]
```diff
@@ -34,14 +34,33 @@ jobs:
         with:
           version: "v0.7.0"
 
-      - name: Run Make
-        run: make
+      # TODO(zzxwill) This step and the follow one need to be removed after rudr admin:init is ready
```
Comment (wonderflow) [inline .github/workflows/go.yml:37]: now it's ready， try using that

Comment (zzxwill) [reply .github/workflows/go.yml:37]: I verified `vela system:init`, and found it just installed oam runtime, but didn't install definition manifests. Is it by-design?

Comment (wonderflow) [reply .github/workflows/go.yml:37]: no, we should fix it

**Pull request merged**